sdk/core/azure-core/inc/response.hpp

```diff
@@ -21,13 +21,16 @@ namespace Azure { namespace Core {
     // Do not give up raw response ownership.
     Http::RawResponse& GetRawResponse() { return *this->m_rawResponse; }
 
-    T& operator=(const Response& other) = delete;
-
     const T* operator->() const { return &this->m_value; };
     T* operator->() { return &this->m_value; };
     T& operator*() { return this->m_value; };
     const T& operator*() const { return this->m_value; };
 
     T ExtractValue() { return std::move(this->m_value); }
+
+    std::unique_ptr<Http::RawResponse> ExtractRawResponse()
```

Comment:
I'm not sure that "Extract" is idiomatic to C++. I'd like @CaseyCarter / @BillyONeal to comment on pattern and what is the best naming.

Reply:
Sure, this will be blocked by @CaseyCarter and @BillyONeal 's comment then.

BTW, if this is not idiomatic for C++, we need to also change the name of `ExtractValue`.

Reply:
I think Extract is fine.

Reply:
I still believe we don't need `ExtracRawResponse`.  I would expect only T to be extracted at one point from responseT to forget about ResposeT (and rawResponse).  But do we expect to extract rawResponse and forget about T?

Reply:
This kinda stuff is why I was recommending literally:

```
template<class T>
struct Response {
    std::unique_ptr<RawResponse> RawResponse;
    T Value;
};
```

i.e. `Response` is just an aggregate with 2 members, akin to `std::pair`.

Alternately consider:

```
template<class T>
struct Response : T { // note derivation from T
    std::unique_ptr<RawResponse> RawResponse;
};
```

Reply:
We have no real idioms<sup>1</sup> for "this function yanks something out of this object and returns it." I suggested `ExtractValue` because the original name `GetValue` suggests that it's returning a copy. "Extract" more clearly implies that the target object is modified by the call.

<sup>1</sup>The closest examples we have in the Standard Library are functions that require rvalue arguments and return rvalue references like
```c++
// "tuple of strings" means tuple<string, /* maybe other things */>; these return string&&
std::get<0>(std::move(some_tuple_of_strings));
std::get<0>(some_function_that_returns_a_tuple_of_strings());
```
or `optional::value()`, which can be called on `const`/non-`const` lvalue/rvalue `optional`s, which returns a `T&&` when called on a non-const rvalue `optional<T>`. There's no consensus even among WG21 whether such functions should return `T` (always extract the `T`) or `T&&` (give the caller the option to extract the `T`), or even whether it's a good idea to have `std::move(something).some_member()` do something other than leave `something` in a garbage state. 

TLDR: I don't consider the Standard Library to provide good precedent here.

Reply:
> 
> 
> This kinda stuff is why I was recommending literally:
> 
> ```
> template<class T>
> struct Response {
>     std::unique_ptr<RawResponse> RawResponse;
>     T Value;
> };
> ```
> 
> i.e. `Response` is just an aggregate with 2 members, akin to `std::pair`.
> 
> Alternately consider:
> 
> ```
> template<class T>
> struct Response : T { // note derivation from T
>     std::unique_ptr<RawResponse> RawResponse;
> };
> ```

if we use struct on the template and expose the fields as public, is it possible to have them mutated?  
Also, I was looking for a way to make ResponseT to be use as a T directly, instead of doing `Response.value.method()` to just doing `Response->method()`

Reply:
> > This is needed because in the convenience layer, some APIs will reconstruct the returned Response object.
> 
> Why do we want to reconstruct the RawResponse? I though we only construct from a RawResponse. Can you elaborate on it?

Sorry, missed this comment. Basically the `Response<T>` is returned by the protocol and convenience layer APIs. Sometimes the returned value is again parsed or reconstructed to another `Response<T>`.

Two examples of the necessity:
1. Some DataLake APIs calls Blob service convenience APIs for interop as a universal strategy among track two SDKs. Blob APIs would return `Response<SomeBlobResult>` and the DataLake API needs to reconstruct it to `Response<SomeDataLakeResult>`.
2. Some protocol layer API has multiple functionalities, e.g. the set blob properties REST API can set the blob properties and metadata, so the protocol layer API `SetPropeties` will be called in `SetProperties` and `SetMetadata` in convenience layer, the `Response<PropertiesInfo>` returned from protocol layer will then need to be changed to `Response<MetadataInfo` for `SetMetadata` APIs.